### PR TITLE
feat: support DPA4 and DPA4C deployment

### PR DESCRIPTION
**Pull request by SchrodingersCattt**
## Summary

- add the DeePMD-kit PyTorch Exportable backend (`pytorch-exportable` / `pt-expt`) for training
- support DPA4 and DPA4C `.pt2` deployment models in LAMMPS exploration
- freeze and optionally compress checkpoints once before exploration tasks fan out
- keep TensorFlow as the default and preserve the existing `.pb` / `.pth` workflows

## Configuration

DPA4 uses `impl: pytorch` with `model_format: pt2`. DPA4C uses `impl: pytorch-exportable`, `model_devi_backend: pytorch-exportable`, and `model_format: pt2`; `dp_compress: true` enables its optional compressed deployment model.

## Tests

- DPA4 PyTorch checkpoint to `.pt2`
- DPA4C PyTorch Exportable freeze and compression
- `.pt2` LAMMPS model-list replacement
- PyTorch Exportable training command

## Summary by CodeRabbit

- **New Features**
  - Added PyTorch Exportable (`pytorch-exportable`/`pt-expt`) training and deployment support.
  - Added `.pt2` model support, automatic model preparation and compression, and configurable student model counts.
  - Added a dedicated model-preparation stage to workflows.

- **Bug Fixes**
  - Added validation for backend compatibility, model formats, and DPA4/DPA4C template settings.
  - Automatically enforces required atom mapping for `.pt2` LAMMPS workflows.
  - Prevents incompatible DPA4 and DPA4C template combinations.

- **Documentation**
  - Documented DPA4/DPA4C configuration, deployment settings, validation, and smoke testing.

### CODE REVIEW

File: dpgen2/op/run_dp_train.py [hunk 1]
```diff
@@ -318,7 +325,7 @@ def clean_before_quit():
                 shutil.copy2("input_v2_compat.json", train_script_name)
 
             # freeze model
-            if impl == "pytorch":
+            if impl in ["pytorch", "pytorch-exportable"]:
```
[the remark this answers is not in the document]
Comment (SchrodingersCattt) [reply dpgen2/op/run_dp_train.py:328]: Fixed in 22ef000: freeze log sections are now written only in the TensorFlow branch, and the PyTorch Exportable test verifies that no freeze section is emitted.

File: dpgen2/op/run_lmp.py [hunk 2]
```diff
@@ -362,11 +433,81 @@ def get_ele_temp(lmp_log_name):
...
     if head is not None:
         freeze_args += " --head %s" % head
-    freeze_cmd = "dp --pt freeze -c %s %s" % (input_model, freeze_args)
+    if backend == "pytorch-exportable" and Path(frozen_model).suffix == ".pt2":
+        freeze_args += " --lower-kind graph"
+    freeze_cmd = "dp %s freeze -c %s %s" % (
+        _MODEL_BACKEND_FLAGS[backend],
+        input_model,
+        freeze_args,
+    )
```
[the remark this answers is not in the document]
Comment (SchrodingersCattt) [reply dpgen2/op/run_lmp.py]: Fixed in 22ef000: freeze and compress now pass argument lists to `run_command` without shell execution; the DPA4 and DPA4C tests assert the argument lists.

**Comment by SchrodingersCattt** [comment]
Follow-up audit in 45b09a0 addresses the analogous deployment boundaries found in DP-GEN: PT2 exploration inputs now ensure `atom_modify map yes` appears before `read_data`/`read_restart`, and workflow submission rejects cross-backend freezing between PyTorch and PyTorch Exportable checkpoints. PT2 export remains on `run_explore_config`, so AOTInductor compilation uses the exploration/deployment resources. Added regression coverage and documented the compatibility requirement.\n\n@coderabbitai review

File: dpgen2/entrypoint/submit.py [hunk 3]
```diff
@@ -472,6 +473,8 @@ def workflow_concurrent_learning(
     train_style = config["train"]["type"]
     explore_style = config["explore"]["type"]
     fp_style = config["fp"]["type"]
+    if train_style in ["dp", "dp-dist"] and explore_style == "lmp":
+        validate_model_backend(train_config["impl"], explore_config)
```
[the remark this answers is not in the document]
Comment (SchrodingersCattt) [reply dpgen2/entrypoint/submit.py]: Fixed in the follow-up commit. The submission path now uses `train_config.get("impl", "tensorflow")`, preserving the documented TensorFlow default for unnormalized configurations that omit `impl`. Added an explicit normalization regression assertion for the omitted-`impl` case.

**Comment by SchrodingersCattt** [comment]
Follow-up from a real DPA4 -> DPA4C distillation run is now included in `f8a9fab` (with pre-commit follow-up `fbf266d`):

- document the backend-specific acceleration paths: DPA4 uses `model.use_compile` / `model.enable_tf32`, while DPA4C (`pytorch-exportable`) uses `training.enable_compile` / `training.enable_tf32` in the DeePMD template referenced by `train.template_script`
- validate DPA4/DPA4C backend, `.pt2` deployment format, and misplaced compile/TF32 keys before workflow creation
- preserve configurable `numb_models` for from-scratch `dp-dist` students
- prevent slice-only success controls from being forwarded to the non-sliced `prepare-models` step (reproduced with pydflow 1.8.132)
- add focused regression coverage

Validation on the remote head: `13 passed` for `tests/entrypoint/test_submit.py` plus `tests/test_prep_run_lmp_config.py`; the focused run/train and run/LAMMPS set also passed locally (`41 passed`).

**Comment by SchrodingersCattt** [comment]
Addressed both current CodeRabbit findings in `6045f3b`:

- corrected the documentation so DPA4 explicitly uses `model.use_compile`, while DPA4C uses `training.enable_compile`
- changed model-family detection to collect all recognized branches and reject templates mixing DPA4 and DPA4C, independent of JSON branch order
- added a mixed `model.model_dict` regression test

Validation: `tests/entrypoint/test_submit.py` plus `tests/test_prep_run_lmp_config.py` -> `14 passed`.

**Comment by SchrodingersCattt** [comment]
A production DPA4C distillation exposed one more deployment boundary: all four students trained successfully, but PT2 model preparation failed only afterwards because the exploration runtime did not expose a working C++ compiler to AOTInductor (`InvalidCxxCompiler`).

Fixed in `3aa4349`:
- PT2 checkpoint export now performs a compiler preflight on the model-deviation resource before freeze; it honors `CXX`, otherwise probes `g++`/`c++`
- missing or non-executable compiler is reported as an actionable, non-retryable configuration error instead of repeatedly launching doomed freeze jobs
- documentation now calls out the AOTInductor compiler requirement and recommends the same preflight before long campaigns
- regression coverage includes missing/default and invalid configured compilers

Validation: focused export/submission tests -> `31 passed`. The distillation Volc runtime now sets and smoke-checks the conda `CC`/`CXX` toolchain explicitly.

**Comment by SchrodingersCattt** [comment]
Correction: the generic compiler-preflight change from `3aa4349` was broader than DPGEN2 should enforce and has been reverted in `6bf4d51`. PT2 export already surfaces the actual AOTInductor compiler error; deployment toolchain setup belongs to the selected runtime/environment rather than a DPGEN2-wide policy. The project-specific Volc runtime still sets `CC`/`CXX` so this workflow can export successfully, without adding extra pipeline checks for all users.

File: dpgen2/op/run_lmp.py [hunk 4]
```diff
@@ -323,6 +396,56 @@ def set_models(lmp_input_name: str, model_names: List[str]):
...
+    if atom_map_index is not None:
+        if atom_map_index > read_index:
+            raise RuntimeError(
+                "PT2 LAMMPS inputs require 'atom_modify map yes' before "
+                "read_data or read_restart"
+            )
+        return
+
+    lmp_input_lines.insert(read_index, "atom_modify        map yes\n")
```
Comment (wanghan-iapcm) [inline dpgen2/op/run_lmp.py]: This inserts at the index of the matching line, which corrupts `&`-continued commands. Run against the PR head:

```
input:
  if "${restart} > 0" then &
      "read_restart dpgen.restart.*" &
  else &
      "read_data conf.lmp"

output:
  if "${restart} > 0" then &
  atom_modify        map yes
      "read_restart dpgen.restart.*" &
  else &
      "read_data conf.lmp"
```

LAMMPS parses that as `if "${restart} > 0" then atom_modify map yes` and dies with "Illegal if command"; the remaining fragments become orphaned commands. Every `run-lmp` task fails.

The reason CI is green is that `test_pt2_enables_atom_map_before_read` uses the **single-line** form:

```python
'if "${restart} > 0" then "read_restart dpgen.restart.*" else "read_data conf.lmp"\n'
```

which is what `lmp_input.py` generates. But `ensure_pt2_atom_map` runs unconditionally on `lmp_input_name` whenever any model is `.pt2`, including inputs from `LmpTemplateTaskGroup` / `CustomizedLmpTemplateTaskGroup` — user-authored templates, where `&` continuations for restart logic are idiomatic. Those are also the users most likely to be on pt2. Walking back past any preceding line whose stripped content ends in `&` before inserting would fix it.

Two smaller ones in the same function, both reproduced:

- Only the **first** read command is protected. `read_data ... clear ... read_data` gets one insertion; `clear` resets `atom_modify` to defaults, so the second box has no map and the pt2 pair style fails at run time.
- The existence check at line 432 requires the literal `map yes`, so an explicit `atom_modify map array` is left in place and the injected `map yes` is appended after it and wins — silently discarding a deliberate choice. Accepting `map\s+(yes|array|hash)` covers it.

File: dpgen2/entrypoint/submit.py [hunk 5]
```diff
@@ -462,6 +463,92 @@ def get_systems_from_data(data, data_prefix=None):
...
+        )
+        model_type = model_type.lower() if isinstance(model_type, str) else model_type
+        descriptor_type = (
+            descriptor_type.lower()
+            if isinstance(descriptor_type, str)
+            else descriptor_type
+        )
+        if descriptor_type == "dpa4c":
+            families.add("dpa4c")
+        if model_type == "dpa4" or descriptor_type in {"dpa4", "sezm"}:
```
Comment (wanghan-iapcm) [inline dpgen2/entrypoint/submit.py]: Family detection misses the standard multitask template shape, so the validation this PR exists to provide is silently skipped for it.

`descriptor.get("type")` is only read when `descriptor` is a dict (line 483), and `_iter_model_sections` never looks inside `model.shared_dict`. But the established multitask idiom in this codebase puts descriptors in `shared_dict` and references them **by name** — `prep_dp_train.py` has resolved `model_dict[...]["descriptor"]` as a string key into `model["shared_dict"]` since #207. Run against head:

```
{"model": {"shared_dict": {"my_desc": {"type": "dpa4c"}},
           "model_dict": {"a": {"descriptor": "my_desc"}}}}
  -> _model_family = None
  -> validate_dpa_training_template(train_backend="pytorch") -> no error
```

A DPA4C multitask config therefore submits with the wrong backend and fails later, in a GPU pod, at `dp --pt freeze`.

Related asymmetry on this line: `dpa4` is detected from `model_type` **or** `descriptor_type`, but `dpa4c` only from `descriptor_type`. So `{"model": {"type": "dpa4c"}}` also returns `None`. Given the DPA4 example in `docs/input.md` uses `model.type`, a user mirroring it for DPA4C lands here.

Separately and much lower priority: because the two `if`s are not exclusive, a single section with `model.type: "dpa4"` and `descriptor.type: "dpa4c"` adds both families and trips the "cannot mix branches" error even though there is only one branch. I could not find a real template written that way, so treat that as a footnote — resolving the family per section, with `dpa4c` taking precedence, fixes it along with the rest.

File: dpgen2/op/run_lmp.py [hunk 6]
```diff
@@ -362,31 +485,144 @@ def get_ele_temp(lmp_log_name):
...
+    return pytorch_model_name_pattern % index
+
+
+def _compressed_model_name(index, model_format):
+    return "model.%03d.compressed.%s" % (index, model_format)
+
+
+def prepare_dp_models(models, config):
+    """Return frozen models, exporting checkpoints once when needed."""
+    backend = _model_backend(config)
```
Comment (wanghan-iapcm) [inline dpgen2/op/run_lmp.py]: This runs before any extension is inspected, which diverges from `RunLmp.execute` (line 190) and from #207, where the backend is resolved **inside** the `.pt` branch — precisely because it is only needed when there is a checkpoint to freeze.

Consequence, reproduced at head:

```
validate_model_backend("tensorflow", {"model_devi_backend": "tensorflow"})  -> passes
prepare_dp_models(["model.000.pb"], cfg)
  -> RuntimeError: Unsupported model-deviation backend 'tensorflow'
```

`validate_model_backend` early-returns whenever the *training* backend is tensorflow, so submit-time validation cannot catch it. `model_devi_backend` is an unconstrained `str` whose doc reads "The DeePMD backend used to freeze models for exploration" and lists no accepted values, while the sibling `train.config.impl` legitimately takes `"tensorflow"` — so mirroring the two is a natural thing to write. The result is a `.pb`-only workflow, which needs no preparation at all, dying on the first exploration iteration.

And because it is a bare `RuntimeError`, the dflow wrapper exits 1, which is indistinguishable from `TransientError` to Argo's retry predicate — so with `retry_on_transient_error` set it retries a permanent misconfiguration.

Moving `backend = _model_backend(config)` inside the `ext == ".pt"` branch fixes all of it and matches the existing precedent. Restricting the argument's accepted values would be a good belt-and-braces addition.

File: dpgen2/superop/prep_run_lmp.py [hunk 7]
```diff
@@ -148,6 +153,10 @@ def _prep_run_lmp(
     run_executor = init_executor(run_config.pop("executor"))
     template_slice_config = run_config.pop("template_slice_config", {})
 
+    prepare_models_config = deepcopy(run_config)
+    prepare_models_config.pop("continue_on_num_success", None)
+    prepare_models_config.pop("continue_on_success_ratio", None)
```
Comment (wanghan-iapcm) [inline dpgen2/superop/prep_run_lmp.py:159]: `continue_on_failed` is the one step-config key not stripped here, and it is the one that matters most for a fan-in step.

It is a real `dflow.Step` kwarg that maps to Argo's `continueOn.failed`. Users routinely set `continue_on_failed: true` on `run_explore_config` because individual LAMMPS tasks are *expected* to fail — and that config is passed straight through to `PrepRunLmp(run_config=run_explore_config)`. Applied to `prepare-models`, a freeze or compress failure is swallowed, the workflow proceeds, and every sliced `run-lmp` task then fails consuming a missing or incomplete `models` artifact. The real cause ends up buried under N downstream failures.

"Continue on failure" has no useful meaning for a single fan-in step that every downstream task depends on, so I would pop it alongside the other two:

```python
prepare_models_config.pop("continue_on_failed", None)
```

The fact that its two siblings *are* stripped is what makes this read as an oversight rather than a choice.

**Review by wanghan-iapcm — CHANGES_REQUESTED** [review]
Substantial piece of work, and the architecture is right: hoisting the freeze out of the per-task loop into a single `prepare-models` step is the correct fix for what #207 left in `RunLmp.execute`. Blocking on four things in the new code, all reproduced by running it rather than read off the diff.

### What this PR gets right, so nobody re-litigates it

- **The `prepare-models` step config is handled correctly.** It strips `continue_on_num_success` and `continue_on_success_ratio` before spreading into a non-sliced step, and derives `prepare_models_config` *after* `template_slice_config` has already been popped from `run_config`. That is exactly what #408 asks for, and it is more careful than the pre-existing `prep-lmp` step three lines above it.
- **`test_non_sliced_model_preparation_ignores_slice_success_controls` is a genuinely good test.** It asserts `prepare-models` has `None` *while* asserting `run-lmp` still has the values. That differential shape survives mutation; a single-sided assertion would not.
- **The step-key change is complete.** I swept `showkey.py`, `dflow_query.py`, `watch.py`, `download_dpgen2_artifacts.py` and `block.py` — all handle the inserted key, and `test_block_cl.py` / `test_dpgen_loop.py` were both updated. Nothing missed.
- **The docs match the schema.** I built a documented-key vs dargs-declaration table for `model_devi_backend`, `model_format`, `dp_compress` and `impl`; every name and default agrees. The "must contain `model.use_compile`" wording is fine too — it says DPGEN2 validates *placements* and explicitly tells the reader to smoke-test, which is accurate.
- **The duplicate test classes are fixed.** `2d823392` removed them; each class name now appears once and all 26 tests collect. The CodeRabbit thread is stale, not outstanding. One casualty worth knowing about: `test_map_after_read_raises` went away with the shadowed copy, so the `atom_map_index > read_index` branch at line 437 now has no test.

### Blocking

Four findings, all inline. Briefly: `ensure_pt2_atom_map` corrupts `&`-continued commands; `_model_family` silently skips validation for multitask templates; `_model_backend` is called before the extension dispatch so a TensorFlow workflow can die in the pod; and `continue_on_failed` is the one step-config key not stripped from the new step.

### Not blocking, but worth a look

**1. `dp-dist` multi-model looks unusable with this repo's own example.** `numb_models` is now declared for `dp-dist`, and `submit.py` raises when it is not 1 *and* `student_model_path` is set — but there is no equivalent check for `student_model_uri`, so that path silently trains N students where it previously forced 1. More importantly, `RunLmp.execute` still has

```python
if teacher_model is not None:
    assert len(model_files) == 1, "One model is enough in knowledge distillation"
```

and `examples/ch4/input_dist.json` — the only `dp-dist` example in the repo — sets `explore.config.teacher_model_path`. I ran it: `numb_models: 4` with that example passes submit-time validation cleanly, and would then assert on every `run-lmp` task. If configurable student counts is a headline feature, that assertion needs to move or relax.

**2. `freeze_model` / `compress_model` raise `TransientError` for deterministic failures.** `run_dp_train.py` already raises `FatalError` for the identical `dp freeze failed`, so this is inconsistent within the repo. It matters more here than it used to: `prepare_dp_models` re-freezes *every* model from scratch on each attempt, so with the usual `numb_models: 4` a single bad checkpoint burns 4x the GPU pod time per retry. Your own `3aa43499` diagnosed exactly this class of failure from a real DPA4C run and used `FatalError`; the revert in `6bf4d515` was about not enforcing toolchain policy, which I agree with — but the error *classification* is a separate question and the revert took it along. `freeze_model`'s choice is pre-existing; `compress_model` is new and copies it.

File: dpgen2/op/run_lmp.py (hunk 4 above)
Comment (SchrodingersCattt) [reply dpgen2/op/run_lmp.py]: Fixed in cd3855b. `ensure_pt2_atom_map` now parses LAMMPS logical commands (including trailing `&` continuations), inserts the map before `create_box`, `read_data`, or `read_restart`, resets after `clear`, accepts `yes`/`array`/`hash`, and retains the late-map validation. Focused regression tests: 7 passed, including 2 subtests.

File: dpgen2/entrypoint/submit.py (hunk 5 above)
Comment (SchrodingersCattt) [reply dpgen2/entrypoint/submit.py]: Fixed in 6dc1c77. Family resolution now follows string references through `model.shared_dict`, recognizes direct `model.type: dpa4c`, applies DPA4C precedence within one model section, and rejects only a real DPA4/DPA4C mix across branches. Focused regression tests: 5 passed.

File: dpgen2/op/run_lmp.py (hunk 6 above)
Comment (SchrodingersCattt) [reply dpgen2/op/run_lmp.py]: Fixed in fe6174a. Backend resolution is now limited to `.pt` checkpoints that actually need freezing; already frozen `.pb`, `.pth`, and `.pt2` models pass through without requiring a PyTorch deployment backend, while unknown extensions still fail explicitly. Focused regression tests: 7 passed.

File: dpgen2/superop/prep_run_lmp.py (hunk 7 above)
Comment (SchrodingersCattt) [reply dpgen2/superop/prep_run_lmp.py:159]: Fixed in 0c532a4. `prepare-models` now strips `continue_on_failed` together with the two success controls, while the sliced `run-lmp` step retains all three configured values. Regression coverage asserts the differing prepare/run behavior; the final changed-file pre-commit suite passes and Linux CI is running.

**Comment by SchrodingersCattt** [comment]
Follow-up fixes for the two non-blocking review items are now included:

- `9effb94`: supports one teacher plus multiple distillation students, keeps the teacher first in the LAMMPS model list, and rejects using a singular student path/URI to initialize multiple students.
- `0afd79b`: classifies deterministic `dp freeze` / `dp compress` failures as `FatalError`; LAMMPS execution failures remain `TransientError`.

All changed files pass pre-commit, changed production files pass Pyright with 0 errors, and focused regression tests pass. Linux CI is running for the full suite.

@coderabbitai review

**Comment by SchrodingersCattt** [comment]
One final ordering edge case from the multi-student audit is fixed in `c595fac`: when `shuffle_models: true`, only student models are shuffled and the teacher remains at model position 0. The regression test forces the two students to reverse and verifies the resulting `pair_style` still starts with the teacher. Focused test: 1 passed; pre-commit and Pyright pass.

@coderabbitai review

File: dpgen2/superop/prep_run_lmp.py (hunk 7 above)
Comment (SchrodingersCattt) [reply dpgen2/superop/prep_run_lmp.py:159]: CI clarified one dflow detail: after the key is removed, `Step.continue_on_failed` materializes as its default `False`, whereas the two optional success thresholds materialize as `None`. Commit 147c7ec corrects only that regression assertion; the implementation still pops `continue_on_failed`, and `run-lmp` still retains the configured `True`. The exact focused test now passes.

**Comment by SchrodingersCattt** [comment]
Final head `147c7ec` is fully green: Python 3.9, Python 3.13, Pyright, pre-commit, docs, Codecov, and CodeRabbit all pass. The four blocking maintainer threads have point-to-point replies and remain open for reviewer confirmation.

**Review by wanghan-iapcm — APPROVED** [review]
All four points are addressed. I checked each one at 147c7ecb by running it rather than taking the commit messages on trust: `ensure_pt2_atom_map` now inserts before a whole logical command, so `&`-continued blocks survive, it does not duplicate an existing map, and it re-inserts after `clear`; family detection follows `shared_dict` string references, so the multitask shape resolves, while a genuine DPA4/DPA4C mix still raises; `prepare_dp_models` checks the suffix first, so already-frozen models never reach backend resolution; and `continue_on_failed` is stripped from prepare-models alongside the two success controls.
